beacon-chain/core/blocks/block_operations.go

```diff
@@ -250,139 +237,114 @@ func ProcessAttesterSlashings(
-			slashableVoteData2Attestation,
+			data1,
+			data2,
 		)
 	}
-
-	// Unless the following holds, the slashing is invalid:
-	// (vote1.justified_slot < vote2.justified_slot) &&
-	// (vote2.justified_slot + 1 == vote2.slot) &&
-	// (vote2.slot < vote1.slot)
-	// OR
-	// vote1.slot == vote2.slot
-	justificationValidity :=
-		(slashableVoteData1Attestation.JustifiedSlot < slashableVoteData2Attestation.JustifiedSlot) &&
-			(slashableVoteData2Attestation.JustifiedSlot+1 == slashableVoteData2Attestation.Slot) &&
-			(slashableVoteData2Attestation.Slot < slashableVoteData1Attestation.Slot)
-
-	slotsEqual := slashableVoteData1Attestation.Slot == slashableVoteData2Attestation.Slot
-
-	if !(justificationValidity || slotsEqual) {
-		return fmt.Errorf(
-			`
-			Expected the following conditions to hold:
-			(slashableVoteData1.JustifiedSlot <
-			slashableVoteData2.JustifiedSlot) &&
-			(slashableVoteData2.JustifiedSlot + 1
-			== slashableVoteData1.Slot) &&
-			(slashableVoteData2.Slot < slashableVoteData1.Slot)
-			OR
-			slashableVoteData1.Slot == slashableVoteData2.Slot
-
-			Instead, received slashableVoteData1.JustifiedSlot %d,
-			slashableVoteData2.JustifiedSlot %d
-			and slashableVoteData1.Slot %d, slashableVoteData2.Slot %d
-			`,
-			slashableVoteData1Attestation.JustifiedSlot,
-			slashableVoteData2Attestation.JustifiedSlot,
-			slashableVoteData1Attestation.Slot,
-			slashableVoteData2Attestation.Slot,
-		)
+	// Verify the slashing is a double vote or a surround vote.
+	if !(isDoubleVote(data1, data2) || isSurroundVote(data1, data2)) {
+		return errors.New("attester slashing is not a double vote nor surround vote")
+	}
+	if err := verifySlashableAttestation(slashableAttestation1, verifySignatures); err != nil {
+		return fmt.Errorf("could not verify attester slashable attestation data 1: %v", err)
+	}
+	if err := verifySlashableAttestation(slashableAttestation2, verifySignatures); err != nil {
+		return fmt.Errorf("could not verify attester slashable attestation data 2: %v", err)
 	}
 	return nil
 }
 
-func attesterSlashingPenalizedIndices(slashing *pb.AttesterSlashing) ([]uint64, error) {
-	indicesIntersection := sliceutil.Intersection(
-		slashing.SlashableVote_1.ValidatorIndices,
-		slashing.SlashableVote_2.ValidatorIndices)
-	if len(indicesIntersection) < 1 {
-		return nil, fmt.Errorf(
-			"expected intersection of vote indices to be non-empty: %v",
-			indicesIntersection,
-		)
+func attesterSlashableIndices(beaconState *pb.BeaconState, slashing *pb.AttesterSlashing) ([]uint64, error) {
+	slashableAttestation1 := slashing.SlashableAttestation_1
+	slashableAttestation2 := slashing.SlashableAttestation_2
+	// Let slashable_indices = [index for index in slashable_attestation_1.validator_indices if
+	//   index in slashable_attestation_2.validator_indices and
+	//   state.validator_registry[index].penalized_epoch > get_current_epoch(state)].
+	var slashableIndices []uint64
+	for _, idx1 := range slashableAttestation1.ValidatorIndices {
```

Comment:
How big are these lists? This is O(n^2) but could be O(n*logn) if we sorted and traversed linearly. 

IMO, only worth optimizing if the lists are non trivial size because we'd want to prove via some benchmark that it is actually faster :)

Reply:
They will be relatively, there is a cap on max slashings for attesters so I wouldn't be too concerned about this